catboost/libs/metrics/metric.cpp

```diff
@@ -466,6 +467,79 @@ void TMultiRMSEMetric::GetBestValue(EMetricBestValue* valueType, float* /*bestVa
     *valueType = EMetricBestValue::Min;
 }
 
+/* MultiRMSEWithMissingValues */
+namespace {
+    struct TMultiRMSEWithMissingValues final: public TAdditiveMultiRegressionMetric {
+        explicit TMultiRMSEWithMissingValues(const TLossParams& params)
+            : TAdditiveMultiRegressionMetric(ELossFunction::MultiRMSE, params)
+        {}
+
+        static TVector<THolder<IMetric>> Create(const TMetricConfig& config);
+
+        TMetricHolder EvalSingleThread(
+            TConstArrayRef<TVector<double>> approx,
+            TConstArrayRef<TVector<double>> approxDelta,
+            TConstArrayRef<TConstArrayRef<float>> target,
+            TConstArrayRef<float> weight,
+            int begin,
+            int end
+        ) const override;
+        void GetBestValue(EMetricBestValue* valueType, float* bestValue) const override;
+        double GetFinalError(const TMetricHolder& error) const override;
+    };
+}
+
+// static
+TVector<THolder<IMetric>> TMultiRMSEWithMissingValues::Create(const TMetricConfig& config) {
+    return AsVector(MakeHolder<TMultiRMSEWithMissingValues>(config.Params));
+}
+
+TMetricHolder TMultiRMSEWithMissingValues::EvalSingleThread(
+    TConstArrayRef<TVector<double>> approx,
+    TConstArrayRef<TVector<double>> approxDelta,
+    TConstArrayRef<TConstArrayRef<float>> target,
+    TConstArrayRef<float> weight,
+    int begin,
+    int end
+) const {
+    const auto evalImpl = [=](bool useWeights, bool hasDelta) {
+        const auto realApprox = [=](int dim, int idx) { return approx[dim][idx] + (hasDelta ? approxDelta[dim][idx] : 0); };
+        const auto realWeight = [=](int idx) { return useWeights ? weight[idx] : 1; };
+        
+        TMetricHolder error(target.size()*2);
+        for (auto dim : xrange(target.size())) {
+            double sumWeights = 0.0;
+            double sumErrors = 0.0;
+            for (auto i : xrange(begin, end)) {
+                if (!std::isnan(target[dim][i])) {
+                    sumErrors += realWeight(i) * Sqr(realApprox(dim, i) - target[dim][i]);
+                    sumWeights += realWeight(i);
+                } else {
+                    continue;
+                }
+            }
+            error.Stats[dim*2] += sumErrors;
+            error.Stats[dim*2+1] += sumWeights;
+        }
+        
+        return error;
+    };
+
+    return DispatchGenericLambda(evalImpl, !weight.empty(), !approxDelta.empty());
+}
+
+double TMultiRMSEWithMissingValues::GetFinalError(const TMetricHolder& error) const {
+    double finalError = 0.0;
+    for (size_t dim = 0; dim < error.Stats.size(); dim+=2) {
+        finalError += error.Stats[dim]/error.Stats[dim+1];
```

Comment:
We should check `error.Stats[dim+1]` for 0

Reply:
Fixed